JPS_BASE_LIB/src/main/java/uk/ac/cam/cares/jps/base/derivation/DerivationSparql.java

```diff
@@ -1578,52 +2348,338 @@ boolean validatePureInputs() {
+	 * @param retrievedInputsAt
+	 */
+	boolean reconnectNewDerivedIRIs(List<TriplePattern> outputTriples,
+			Map<String, List<String>> newIriDownstreamDerivationMap, String derivation,
+			Long retrievedInputsAt) {
+		ModifyQuery modify = Queries.MODIFY();
+		SubSelect sub = GraphPatterns.select();
+
+		// insert all output triples
+		outputTriples.forEach(t -> modify.insert(t));
+
+		// insert triples of new derived IRI and the derivations that it should be
+		// connected to
+		newIriDownstreamDerivationMap.forEach((newIRI, downstreamDerivations) -> {
+			// add <newIRI> <belongsTo> <derivation> for each newIRI
+			modify.insert(iri(newIRI).has(belongsTo, iri(derivation)));
+			// if this <newIRI> is inputs to other derivations, add
+			// <downstreamDerivation> <isDerivedFrom> <newIRI>
+			if (!downstreamDerivations.isEmpty()) {
+				downstreamDerivations.stream().forEach(dd -> modify.insert(iri(dd).has(isDerivedFrom, iri(newIRI))));
+			}
+		});
+
+		// create variables for sub query
+		Variable d = SparqlBuilder.var("d"); // this derivation
+		Variable ups = SparqlBuilder.var("ups"); // upstream inputs/derivations
+
+		// variables for old outputs (entities)
+		Variable e = SparqlBuilder.var("e");
+		Variable p1 = SparqlBuilder.var("p1");
+		Variable o = SparqlBuilder.var("o");
+		Variable s = SparqlBuilder.var("s");
+		Variable p2 = SparqlBuilder.var("p2");
+
+		// variables for timestamp
+		Variable unixtimeIRI = SparqlBuilder.var("timeIRI");
+		Variable dTs = SparqlBuilder.var("dTs"); // timestamp of this derivation
+		Variable upsTs = SparqlBuilder.var("upsTs"); // timestamp of upstream
+
+		// variables for status related concepts
+		Variable status = SparqlBuilder.var("status");
+		Variable statusType = SparqlBuilder.var("statusType");
+		Variable sndIRI = SparqlBuilder.var("sndIRI");
+
+		// filter ?derivationTimestamp < ?upstreamTimestamp
+		Expression<?> tsFilter = Expressions.lt(dTs, upsTs);
+
+		// this GraphPattern ensures ?d only exist if it's outdated, otherwise the
+		// SPARQL update will not execute
+		GraphPattern derivationPattern = GraphPatterns
+				.and(new ValuesPattern(d, Arrays.asList(iri(derivation))),
+						d.has(PropertyPaths.path(hasTime, inTimePosition), unixtimeIRI),
+						unixtimeIRI.has(numericPosition, dTs),
+						d.has(PropertyPaths.path(isDerivedFrom, zeroOrOne(belongsTo)), ups),
+						ups.has(PropertyPaths.path(hasTime, inTimePosition, numericPosition), upsTs))
+				.filter(tsFilter);
+		// this GraphPattern queries all the old outputs (entities)
+		GraphPattern entityPattern = GraphPatterns.and(e.has(belongsTo, d), e.has(p1, o), s.has(p2, e).optional());
+		// these patterns query the status of this derivation if it exist, thus optional
+		TriplePattern tp1 = d.has(hasStatus, status);
+		TriplePattern tp2 = status.isA(statusType);
+		TriplePattern tp3 = status.has(hasNewDerivedIRI, sndIRI);
+		GraphPattern statusQueryPattern = GraphPatterns.and(tp1, tp2, tp3.optional()).optional();
```

Comment:
```suggestion
		GraphPattern statusQueryPattern = GraphPatterns.and(tp1, tp2, tp3).optional();
```

Reply:
I think we should remove `TriplePattern tp3 = status.has(hasNewDerivedIRI, sndIRI);` as it is a triple that would appear for asynchronous derivation only, NOT for synchronous derivation. As here we are only calling reconnectNewDerivedIRI when updating synchronous derivations, we can just ignore that.

Reply:
This is addressed in https://github.com/cambridge-cares/TheWorldAvatar/pull/185/commits/b77eb21d7abc361942cf2d20bf4bbc233bb41ca3.